pretyping/unification.ml

```diff
@@ -2380,13 +2380,44 @@ and make_array sigma v =
 
 end
 
+type head_kind =
+| HeadSort
+| HeadProd
+| HeadInd
+| HeadOther
+
+let get_head_kind ~metas env sigma c =
+  let metas n =
+    try Some (Meta.meta_ftype metas n).rebus
+    with Not_found -> None
+  in
+  match Retyping.get_type_of ~metas env sigma c with
+  | ty ->
+    let hd, _ = decompose_app sigma ty in
+    match EConstr.kind sigma hd with
+    | Prod _ -> HeadProd
+    | Sort _ -> HeadSort
+    | Ind _ -> HeadInd
+    | _ -> HeadOther
+  | exception _ -> HeadOther
+
+(* can c have a type of the same shape as knd? *)
+let fast_head_check sigma knd c = match EConstr.kind sigma c, knd with
+| Lambda _, (HeadInd | HeadSort) -> false
+| Sort _, (HeadInd | HeadProd) -> false
+| Construct _, (HeadProd | HeadSort) -> false
+| Prod _, (HeadInd | HeadProd) -> false
+| Ind _, HeadInd -> false
+| _ -> true
```

Comment:
Would it make sense to add cases for primitive types such as int, float, array, and string here?